routes/view_router.js

```diff
@@ -1,101 +1,136 @@
+		'gamers': {query: {username: req.params.username, platform: req.params.platform}}
+	};
+
+	let promises = Object.keys(views).map((key) => queryView(key, views[key].query));
+
+	Bluebird.all(promises).then(async (arrData) => {
+		Object.keys(views).forEach((key, index) => views[key].data = arrData[index]);
+
+		let gamer = sanitizeGamer(views['player_stat_summary'].data[0]);
+		let teammates = sanitizeTeammates(views['teammate_analysis'].data);
+		let grades = views['gamer_stats_graded'].data[0];
+
+		let gamerData = views['gamers'].data[0];
+		let gamerPromise = Bluebird.resolve(gamerData);
+
+		if (!gamerData.needs_update) {
+			gamerData.needs_update = true;
+			gamerPromise = updateGamer({username: gamerData.username, platform: gamerData.platform}, gamerData);
+		}
+
+		await gamerPromise;
+
+		let titleKeys = Object.keys(teammates[0]).filter((key) => !FILTER_KEYS.includes(key)).map(columnToDisplayName);
+
+		let seoMetadata = {
+			title: 'Warzone stats for ' + gamer.username,
+			keywords: ['warzone', 'stats', 'kdr', 'gulag wins'],
+			description: 'KDR: ' + gamer.kdr + ' Gulag Win Rate: ' + gamer.gulag_win_rate
+		};
+
+		res.render('gamer/detail', {
+			gamer: gamer,
+			grades: grades,
+			titleKeys: titleKeys,
+			teammateData: teammates,
+			seoMetadata: seoMetadata,
+			filterKeys: FILTER_KEYS
+		});
+	});
 });
 
+
+
 view_router.get('/gamers', async (req, res) => {
-    let viewName = 'player_stat_summary';
-    let queryParams = req.query;
-
-    let submittedUsername = queryParams.submittedUsername || '';
-    let submissionError = queryParams.submissionError || '';
-
-    let data = await queryView(viewName);
-    let mappedGamers = data.map(mapGamer);
-    let seoMetadata = {
-        title: 'Warzone Gamers',
-        keywords: ['warzone', 'stats', 'kdr', 'gulag wins'],
-        description: 'Warzone stats for ' + data.length + ' gamers'
-    };
-    res.render('gamer/list', {
-        gamers: mappedGamers,
-        seoMetadata: seoMetadata,
-        submittedUsername: submittedUsername,
-        submissionError: submissionError,
-        filterKeys: FILTER_KEYS
-    });
+	let viewName = 'player_stat_summary';
+	let queryParams = req.query;
+
+	let submittedUsername = UtilityService.validateItem(queryParams.submittedUsername, 'string', '');
+	let submissionError = UtilityService.validateItem(queryParams.submissionError, 'string', '');
+
+	let rawGamerList = await queryView(viewName);
+	let sanitizedGamers = rawGamerList.map(sanitizeGamer);
+
+	let seoMetadata = {
+		title: 'Warzone Gamers',
+		keywords: ['warzone', 'stats', 'kdr', 'gulag wins'],
+		description: 'Warzone stats for ' + sanitizedGamers.length + ' gamers'
+	};
+
+	res.render('gamer/list', {
+		gamers: sanitizedGamers,
+		seoMetadata: seoMetadata,
+		submittedUsername: submittedUsername,
+		submissionError: submissionError,
+		filterKeys: FILTER_KEYS
+	});
 });
 
 
+//===---==--=-=--==---===----===---==--=-=--==---===----//
+//PRIVATE METHODS
```

Comment:
These should probably be moved into some sort of utility file. I don't like this extra logic in the `routes/view_router.js` 

`view_router.js` ideally should only have the routing logic. 

Maybe we need a `renderer.js` or something I have that on my EJS website.

Reply:
Absolutely, this should probably be done in a Gamer component. My big idea was moving towards a component-structure, like this https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/projectstructre/breakintcomponents.md